ddtrace/tracer/sampler.go

```diff
@@ -147,3 +156,261 @@ func (ps *prioritySampler) apply(spn *span) {
+		return rate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", rate, r)
+	return rate
+}
+
+func newRateLimiter(r float64) *rate.Limiter {
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return nil
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return nil
+	}
+	switch {
+	case l < 0.0:
+		return nil
+	case l == 0.0:
+		return rate.NewLimiter(0.0, 0)
+	case (r * l) < 1.0:
+		return rate.NewLimiter(rate.Limit(l), 1)
+	default:
+		return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+	}
+}
+
+func (rs *rulesSampler) apply(span *span) bool {
+	matched := false
+	sr := 0.0
+	for _, v := range rs.rules {
+		if v.match(span) {
+			matched = true
+			sr = v.Rate
+			break
+		}
+	}
+	if !matched {
+		return false
+	}
+	// rate sample
+	span.SetTag("_dd.rule_psr", sr)
+	if !sampledByRate(span.TraceID, sr) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	// global rate limit and effective rate calculations
+	defer rs.mu.Unlock()
+	rs.mu.Lock()
+	if ts := time.Now().Unix(); ts > rs.ts {
+		// update "previous rate" and reset
+		if ts-rs.ts == 1 && rs.total > 0 && rs.allowed > 0 {
+			rs.previousRate = float64(rs.allowed) / float64(rs.total)
+		} else {
+			rs.previousRate = 0.0
+		}
+		rs.ts = ts
+		rs.allowed = 0
+		rs.total = 0
+	}
+
+	rs.total++
+	if rs.limiter != nil && !rs.limiter.Allow() {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+	} else {
+		rs.allowed++
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoKeep)
+	}
+	// calculate effective rate, and tag the span
+	er := (rs.previousRate + (float64(rs.allowed) / float64(rs.total))) / 2.0
+	span.SetTag("_dd.limit_psr", er)
+
+	return true
+}
+
+// SamplingRule is used for applying sampling rates to spans that match
+// the service name, operation or both.
+// It's recommended to use the helper functions (ServiceRule, OperationRule,
+// ServiceOperationRule) instead of directly creating a SamplingRule.
+type SamplingRule struct {
+	Service   *regexp.Regexp
+	Operation *regexp.Regexp
+	Rate      float64
+	err       error
+}
+
+// ServiceRule returns a *SamplingRule that applies the provided sampling rate
+// to spans that match the service name provided.
+// The value for service can include regular expression syntax.
+func ServiceRule(service string, rate float64) *SamplingRule {
+	re, err := regexp.Compile(anchoredRE(service))
+	if err != nil {
+		err = fmt.Errorf("service '%s' is invalid: %v", service, err)
+	}
+	return &SamplingRule{
+		Service: re,
+		Rate:    rate,
+		err:     err,
+	}
+}
+
+// OperationRule returns a *SamplingRule that applies the provided sampling rate
+// to spans that match the operation name provided.
+// The value for operation can include regular expression syntax.
```

Comment:
I would like to add some more thoughts that came to me while pondering on reducing the complexity of the implementation and increasing the simplicity of the API.

I want to re-emphasise once more that we should do as initially planned in https://github.com/DataDog/dd-trace-go/pull/539#issuecomment-559706267: have the helper functions be an exact match, which is what most of the users will want, and leave regular expressions exclusively to using the structure directly.

Instead of adding prefixes, suffixes and escaping strings to convert them into "exact match" regular expressions, which seems hacky, complex and error prone, I would suggest adding two more un-exported fields to the `SamplingRule` structure, perhaps something like: `exactService, exactOp string` which will be simply matched against with the equals (`==`) operator when set. The advantages of this (if not already obvious) are that the implementation becomes dead simple and no hacky regexp parsing is required, and most of all, the performance increases considerably because now we will do a strict equals-operator comparison, as opposed to a regexp match.

When used in practice, we should keep in mind that this is going to get called a lot and performance is a concern here. Adding the exact match strings (which is what most users will use) should have a great positive impact.

Reply:
Should the rules loaded from the JSON-encoded env variable all be exact match?